templates/rust/macros.tera

```diff
@@ -145,11 +142,24 @@ pub fn {{cluster_func}}(self) -> [{{cluster_struct_path}};{{cluster.dim}}] {
 {%- set cluster_struct = cluster.struct_id | to_struct_id -%}
 {%- set cluster_mod = cluster.module_id -%}
 #[doc = "{{cluster.description | svd_description_to_doc}}"]
-#[derive(Copy, Clone, Eq, PartialEq)]
-pub struct {{ cluster_struct }}{pub(crate) ptr: *mut u8}
-unsafe impl ::core::marker::Send for {{ cluster_struct}} {}
-unsafe impl ::core::marker::Sync for {{ cluster_struct }} {}
-impl {{cluster_struct}} {
+#[non_exhaustive]
+pub struct _{{ cluster_struct }};
+
+#[doc = "{{cluster.description | svd_description_to_doc}}"]
+pub type {{ cluster_struct }} = &'static _{{ cluster_struct }};
+
+unsafe impl ::core::marker::Sync for _{{ cluster_struct }} {}
+impl _{{cluster_struct}} {
+    #[allow(unused)]
+    pub(crate) unsafe fn from_ptr(ptr: *mut u8) -> &'static Self {
```

Comment:
`from_ptr` and `as_ptr` should be moved in a trait named e.g. `PtrT` and then add `impl PtrT` for _Cluster_struct.
This to avoid any conflict with register or cluster name.
The probability of a name conflict is low but it is already happened this kind of conflicts.

Reply:
The disadvantage to this is that then those methods can't be made `const`, at least for now.

How about prefixing the methods, like `_svd2pac_from_ptr`?

Reply:
I don't understand why the methods should be const to declare them in a trait and then implement for each _{{cluster_struct}}. Your proposed solution is acceptable but I think that following one is more elegant:
```rust
trait Getter {
    fn get_register(&self) -> u32;
}

struct MyType {}
impl MyType {
    pub fn get_register(&self)->u32 {
        24
    }
}

impl Getter for MyType {
    fn get_register(&self) -> u32 {
        56
    }
}


fn main() {
    let a = MyType{};
    let b = a.get_register(); //Call inherent implementation
    let c = <MyType as Getter>::get_register(&a); // Call trait implementation
}
```

Reply:
What I meant is that currently `const` methods in traits are unsupported, so `from_ptr` and `as_ptr` could not be made `const`. This means many common register access functions (e.g. [here](https://github.com/Infineon/svd2pac/blob/b3ca2700015bce0f02752a392d8b16987886b1c5/templates/rust/macros.tera#L23)) would have to be made non-`const` if we made `from_ptr` and `as_ptr` trait methods.

I don't think it's a big deal, but it would be an externally-visible change to the code, whereas how `from_ptr` and `as_ptr` are implemented are private implementation details that aren't visible to users.

Reply:
I understood I was looking at `from_ptr` that is not `const` and I overlooked `as_ptr` that is not const.
I agree with you that is not a big deal not having `const` function. In my experience if #[inline(always)] is added and it used any optimization level except `opt-level = 0` there is no runtime cost thanks to constant propagation optimization.

Regarding the externally visible change we can declare the trait `pub (crate)`.